File: rules/java/tests/test_idor_tn.java
// True negative: scoped to current user (safe)
@RestController
public class InvoiceController {

    @Autowired
    private InvoiceRepository invoiceRepo;

    @GetMapping("/invoice/{id}")
    // ok: ez-spring-idor-path-variable
    public Invoice getInvoice(@PathVariable Long id, Principal principal) {
        Invoice invoice = invoiceRepo.findById(id).orElseThrow();
        if (invoice.getOwnerId().equals(principal.getName())) {
            return invoice;
        }
        throw new AccessDeniedException("Not your invoice");
    }
}

File: rules/python/tests/test_idor_tn.py
# True negative: lookup scoped to current user (safe)
from django.http import JsonResponse

def get_invoice(request):
    # ok: ez-django-idor-get-parameter
    invoice_id = request.GET.get("id")
    invoice = Invoice.objects.filter(user=request.user).get(pk=invoice_id)
    return JsonResponse({"total": invoice.total})
